flow/record/base.py

```diff
@@ -661,7 +691,56 @@ def DynamicDescriptor(name, fields):
     return RecordDescriptor(name, [("dynamic", field) for field in fields])
 
 
-def open_path(path, mode, clobber=True):
+def open_stream(fp: BinaryIO, mode: str) -> BinaryIO:
+    if not hasattr(fp, "peek"):
+        fp = Peekable(fp)
+
+    # We peek into the file at the maximum possible length we might need, which is the amount of bytes needed to
+    # determine whether a stream is a RECORDSTREAM or not.
+    peek_data = fp.peek(RECORDSTREAM_MAGIC_DEPTH)
+
+    # If the data stream is compressed, we wrap the file pointer in a reader that can decompress accordingly.
+    if peek_data[:2] == GZIP_MAGIC:
+        fp = gzip.GzipFile(fileobj=fp, mode=mode)
+    elif HAS_BZ2 and peek_data[:3] == BZ2_MAGIC:
+        fp = bz2.BZ2File(fp, mode=mode)
+    elif HAS_LZ4 and peek_data[:4] == LZ4_MAGIC:
+        fp = lz4.open(fp, mode=mode)
+    elif HAS_ZSTD and peek_data[:4] == ZSTD_MAGIC:
+        dctx = zstd.ZstdDecompressor()
+        fp = dctx.stream_reader(fp)
+
+    return fp
+
+
+def find_adapter_for_stream(fp: BinaryIO) -> tuple[BinaryIO, str]:
+    new_stream = Peekable(fp)
+    peek_data = new_stream.peek(RECORDSTREAM_MAGIC_DEPTH)
+    if new_stream.peeked_data is not None:
+        peek_data = new_stream.peeked_data
```

Comment:
Why not just call `.peek` again?

Reply:
As calling `.peek` on the same `Peekable` instance isn't allowed, I initially opted for the `peeked_data` variable. However, this is unnecessarily complex, so I opted to just always wrap the incoming `fp` within a new `Peekable` instance for the `find_adapter_for_stream` function.